### PR TITLE
feat: cascade delete agent

### PR DESCRIPTION
**Pull request by 0xbbjoker**
# Relates to

Database relationship management and agent cleanup functionality

# Risks

**Medium** - Changes database deletion behavior which could affect data integrity and related records

# Background

## What does this PR do?

This PR implements cascade delete functionality for agents in the ElizaOS database. When an agent is deleted, all related records (such as messages, memories, relationships, goals, and other associated data) are automatically deleted to maintain database integrity and prevent orphaned records.

Key changes:
- **Database schema updates**: Added cascade delete constraints to foreign key relationships
- **Agent deletion logic**: Enhanced agent removal to properly clean up all dependent data
- **Data integrity**: Ensures no orphaned records remain when agents are removed
- **Cleanup automation**: Eliminates need for manual cleanup of related agent data

## What kind of change is this?

**Features** (non-breaking change which adds functionality)

## Why are we doing this? Any context or related work?

When agents are deleted from the system, previously their related data (messages, memories, goals, etc.) would remain in the database as orphaned records. This creates:
- Database bloat with unnecessary data
- Potential data integrity issues
- Difficulty in completely removing agent traces
- Inefficient database performance over time

Cascade delete functionality ensures complete and clean agent removal while maintaining referential integrity.

# Documentation changes needed?

**My changes require a change to the project documentation.**
- Database schema documentation needs updates to reflect new cascade relationships
- Agent management documentation should include information about cascade deletion behavior
- Migration notes for existing installations

# Testing

## Where should a reviewer start?

1. Review database schema changes and foreign key constraint modifications
2. Test agent deletion functionality with existing agents that have associated data
3. Verify that all related records are properly removed when an agent is deleted
4. Check that normal agent operations are not affected by the schema changes

## Detailed testing steps

- Create a test agent with associated data (messages, memories, goals, relationships)
- Verify the agent and all related data exists in the database
- Delete the agent using the standard deletion method
- Confirm that:
  - The agent record is deleted
  - All related messages are deleted
  - All memories associated with the agent are deleted
  - All goals and relationships are deleted
  - No orphaned records remain in any related tables
- Test edge cases:
  - Deleting agents with large amounts of associated data
  - Concurrent operations during agent deletion
  - Rollback behavior if deletion fails partway through

## Database changes

- Added `ON DELETE CASCADE` constraints to foreign key relationships pointing to agent records
- Updated foreign key constraints on tables: messages, memories, goals, relationships, and other agent-related tables
- Modified database migration scripts to handle existing data during constraint updates

## Deployment instructions

**⚠️ Important**: This change modifies database constraints. For existing installations:

1. Backup your database before applying this update
2. Run the migration script to update foreign key constraints
3. Test agent deletion functionality in a staging environment first
4. Monitor database performance after deployment

## Discord username

0xbbjoker

## Summary by CodeRabbit

- **Bug Fixes**
	- Improved reliability of agent deletion by ensuring all related data (such as memories, logs, tasks, and cache entries) are automatically removed when an agent is deleted.
- **Tests**
	- Added comprehensive integration tests to verify cascade deletion of agent-related data and correct behavior when deleting agents with or without related data.
- **Refactor**
	- Simplified agent deletion logic for better maintainability and performance.
- **Chores**
	- Updated database schema to enforce cascade delete rules on related tables.

### CODE REVIEW

File: packages/plugin-sql/src/schema/serverAgent.ts [hunk 1]
```diff
@@ -1,14 +1,17 @@
...
   'server_agents',
   {
     serverId: text('server_id')
       .notNull()
       .references(() => messageServerTable.id, { onDelete: 'cascade' }),
-    agentId: text('agent_id').notNull(), // This is the agent's UUID
+    agentId: text('agent_id')
+      .notNull()
+      .references(() => agentTable.id, { onDelete: 'cascade' }),
   },
```
[the remark this answers is not in the document]
Comment (0xbbjoker) [reply packages/plugin-sql/src/schema/serverAgent.ts:15]: don't wanna mess with this, will leave as it was.

File: packages/plugin-sql/src/__tests__/integration/agent.test.ts [hunk 2]
```diff
@@ -700,13 +700,198 @@ describe('Agent Integration Tests', () => {
...
+          // Create tasks
+          const taskId = await cascadeAdapter.createTask({
+            id: uuidv4() as UUID,
+            name: 'Test Task',
+            description: 'A test task',
+            roomId: roomId1,
+            worldId: worldId,
+            tags: ['test'],
+            metadata: { priority: 'high' },
+          });
```
[the remark this answers is not in the document]
Comment (0xbbjoker) [reply packages/plugin-sql/src/__tests__/integration/agent.test.ts:829]: the problem here is that task doesn't have the agentId on type so we add it in base.ts actually.
